### DISCUSSION TITLE
Proposal to move some modules out of the main repository

### DISCUSSION THREAD

**Discussion by nobodyiam** [apolloconfig/apollo#4427] [Ideas]
Currently, the [main repository](https://github.com/apolloconfig/apollo) contains many modules. Hence, its releasing process is complex and takes a long time to release, which could be a problem in some situations. For example, if there is a bug in [apollo-client](https://github.com/apolloconfig/apollo/tree/master/apollo-client) and we need to release a patch to fix it, currently we have to also release all the apollo server packages and docker images.

My proposal is to keep the [main repository](https://github.com/apolloconfig/apollo) only contain apollo server packages and move other modules to be standalone repositories under [apolloconfig](https://github.com/apolloconfig).

The details are as below:
* move [apollo-client](https://github.com/apolloconfig/apollo/tree/master/apollo-client) and [apollo-core](https://github.com/apolloconfig/apollo/tree/master/apollo-core) to a standalone repository, could be named as apolloconfig/apollo-java?
* move [apollo-client-config-data](https://github.com/apolloconfig/apollo/tree/master/apollo-client-config-data) to a standalone repository apolloconfig/apollo-client-config-data
* move [apollo-demo](https://github.com/apolloconfig/apollo/tree/master/apollo-demo) to a standalone repository apolloconfig/apollo-demo
* move [apollo-mockserver](https://github.com/apolloconfig/apollo/tree/master/apollo-mockserver) to a standalone repository apolloconfig/apollo-mockserver
* move [apollo-openapi](https://github.com/apolloconfig/apollo/tree/master/apollo-openapi) to a standalone repository apolloconfig/apollo-openapi
* move [apollo-on-kubernetes](https://github.com/apolloconfig/apollo/tree/master/scripts/apollo-on-kubernetes) to a standalone repository apolloconfig/apollo-on-kubernetes
* merge [docker-quick-start](https://github.com/apolloconfig/apollo/tree/master/scripts/docker-quick-start) with the existing [apollo-build-scripts](https://github.com/apolloconfig/apollo-build-scripts/) repository and rename it to apolloconfig/docker-quick-start

With these structural changes, the main repository will look cleaner and each standalone repository could have its own releasing process and schedule.
There may be situations that some feature might introduce changes to multiple repositories which involves extra effort, but I think it's acceptable as these cases are rare.

Please feel free to share your ideas.

**Comment by nobodyiam** [comment]
#4418 is a pull request to introduce the spring cloud apollo client, which is an example that we could use to create as a standalone repository.

**Comment by qmwu2000** [comment]
I think apollo-core is just a module project, it doesn't have any direct user scenarios, so its not standalone project and therefore repository, maybe it's an an part of apollo server, or duplicated in other projects. Am I right?

发自我的iPhone

**Comment by lepdou** [comment]
I agree it. It will be more cleaner.

Currently, there are few changes in apollo-core, and the apollo-client delivery relies on apollo-core. So user's application will directly rely on apollo-core. So it seems more appropriate to put the apollo-core and the apollo-client together. Keep apollo-core and apollo-client has the same version will be more friendly to users.

Then, Apollo server relies on the apollo-core jar package.

> duplicated in other projects

I thank it's may not a good idea.

**Reply by Anilople** [reply]
It seems that apollo-core use in multiple modules.

When run 
```bash
for module in $(find . -maxdepth 1 -name "apollo*" -type d); do
    echo "${module}"
    filenames=$(find ${module} -name "*.java" -type f)
    if [ -n "${filenames}" ]; then
        echo "${filenames}" | xargs grep -h "import com.ctrip.framework.apollo.core" | sort -u
        echo ""
    fi
done
```
in the repo, I found some usage of apollo-core in other modules

```
./apollo-adminservice
import com.ctrip.framework.apollo.core.ConfigConsts;
import com.ctrip.framework.apollo.core.enums.ConfigFileFormat;
import com.ctrip.framework.apollo.core.utils.StringUtils;

./apollo-assembly

./apollo-biz
import com.ctrip.framework.apollo.core.ConfigConsts;
import com.ctrip.framework.apollo.core.enums.ConfigFileFormat;
import com.ctrip.framework.apollo.core.utils.ApolloThreadFactory;
import com.ctrip.framework.apollo.core.utils.StringUtils;

./apollo-buildtools

./apollo-client
import com.ctrip.framework.apollo.core.ApolloClientSystemConsts;
import com.ctrip.framework.apollo.core.ConfigConsts;
import com.ctrip.framework.apollo.core.MetaDomainConsts;
import com.ctrip.framework.apollo.core.ServiceNameConsts;
import com.ctrip.framework.apollo.core.dto.ApolloConfig;
import com.ctrip.framework.apollo.core.dto.ApolloConfigNotification;
import com.ctrip.framework.apollo.core.dto.ApolloNotificationMessages;
import com.ctrip.framework.apollo.core.dto.ServiceDTO;
import com.ctrip.framework.apollo.core.enums.ConfigFileFormat;
import com.ctrip.framework.apollo.core.enums.Env;
import com.ctrip.framework.apollo.core.enums.EnvUtils;
import com.ctrip.framework.apollo.core.schedule.ExponentialSchedulePolicy;
import com.ctrip.framework.apollo.core.schedule.SchedulePolicy;
import com.ctrip.framework.apollo.core.signature.Signature;
import com.ctrip.framework.apollo.core.spi.MetaServerProvider;
import com.ctrip.framework.apollo.core.spi.Ordered;
import com.ctrip.framework.apollo.core.utils.ApolloThreadFactory;
import com.ctrip.framework.apollo.core.utils.ClassLoaderUtil;
import com.ctrip.framework.apollo.core.utils.DeferredLogger;
import com.ctrip.framework.apollo.core.utils.DeferredLoggerFactory;
import com.ctrip.framework.apollo.core.utils.DeprecatedPropertyNotifyUtil;
import com.ctrip.framework.apollo.core.utils.PropertiesUtil;
import com.ctrip.framework.apollo.core.utils.StringUtils;

./apollo-client-config-data
import com.ctrip.framework.apollo.core.ApolloClientSystemConsts;
import com.ctrip.framework.apollo.core.ConfigConsts;
import com.ctrip.framework.apollo.core.spi.Ordered;
import com.ctrip.framework.apollo.core.utils.DeferredLogger;

./apollo-common
import com.ctrip.framework.apollo.core.enums.ConfigFileFormat;
import com.ctrip.framework.apollo.core.enums.Env;
import com.ctrip.framework.apollo.core.enums.EnvUtils;
import com.ctrip.framework.apollo.core.utils.ApolloThreadFactory;
import com.ctrip.framework.apollo.core.utils.ByteUtil;
import com.ctrip.framework.apollo.core.utils.MachineUtil;
import com.ctrip.framework.apollo.core.utils.StringUtils;

./apollo-configservice
import com.ctrip.framework.apollo.core.ConfigConsts;
import com.ctrip.framework.apollo.core.ServiceNameConsts;
import com.ctrip.framework.apollo.core.dto.ApolloConfig;
import com.ctrip.framework.apollo.core.dto.ApolloConfigNotification;
import com.ctrip.framework.apollo.core.dto.ApolloNotificationMessages;
import com.ctrip.framework.apollo.core.dto.ServiceDTO;
import com.ctrip.framework.apollo.core.signature.Signature;
import com.ctrip.framework.apollo.core.utils.ApolloThreadFactory;
import com.ctrip.framework.apollo.core.utils.PropertiesUtil;
import com.ctrip.framework.apollo.core.utils.StringUtils;

./apollo-demo
import com.ctrip.framework.apollo.core.ConfigConsts;
import com.ctrip.framework.apollo.core.enums.ConfigFileFormat;

./apollo-mockserver
import com.ctrip.framework.apollo.core.ApolloClientSystemConsts;
import com.ctrip.framework.apollo.core.dto.ApolloConfig;
import com.ctrip.framework.apollo.core.dto.ApolloConfigNotification;
import com.ctrip.framework.apollo.core.utils.ResourceUtils;

./apollo-openapi
import com.ctrip.framework.apollo.core.ConfigConsts;
import com.ctrip.framework.apollo.core.enums.ConfigFileFormat;

./apollo-portal
import com.ctrip.framework.apollo.core.ConfigConsts;
import com.ctrip.framework.apollo.core.dto.ServiceDTO;
import com.ctrip.framework.apollo.core.enums.ConfigFileFormat;
import com.ctrip.framework.apollo.core.utils.ApolloThreadFactory;
import com.ctrip.framework.apollo.core.utils.NetUtil;
import com.ctrip.framework.apollo.core.utils.PropertiesUtil;
import com.ctrip.framework.apollo.core.utils.ResourceUtils;
import com.ctrip.framework.apollo.core.utils.StringUtils;
```

Whatever the module changed or moved, there is always a common module `xxx` use by apollo-client and apollo-configservice, same situation with openapi and portal.

How we deal with this common module `xxx`? Put it to the client side's git repo, or server side's git repo? It maybe a hard choice.

Or there is another better way to handle it?

**Comment by Anilople** [comment]
apollo's client will be integration with multiple runtime, like 

* raw jdk
* spring
* spring boot
* spring cloud

Is it better split the integration solution to differect maven module?

For example, if user's runtime is raw jdk, user only use raw jdk's maven dependency without care about the spring.

if user's runtime is spring cloud, user choose spring cloud's maven dependency.

shortcoming: user maybe don't know how to choose a dependency because there are multiple dependency.

**Comment by nobodyiam** [comment]
1. Regarding the `apollo-core` module

Since it's only a common module hosting some libs, it's not suitable to split it as a standalone repository. I think it's better to host it with `apollo-client` in a repository named `apollo-java`, as apollo-client is more integrated with it than other modules.
Then apollo server modules could reference it as normal maven dependencies as `apollo-core` is deployed into maven central.
Another reason is that it's more lightweight to release the client libraries than the apollo server packages, so if there is some situation where we need to modify the `apollo-core` module to support features in apollo servers, we could do that without too much effort.

2. Regarding the multiple runtime support like spring boot and spring cloud

Currently `apollo-client` provides support to raw java, spring, and spring boot, so if there is a spring cloud support feature like #4418, I think it's better to treat it as a standalone maven module like `spring-cloud-starter-apollo` so that it's easy to handle the dependencies and provider better integration support. Likewise, we may also have `spring-boot-starter-apollo` in the future, if we find situations that we could better integrate apollo with spring boot.

3. Regarding whether to host client libraries in different standalone repositories or different maven modules in one repository

Apollo also has some other client libraries like `apollo-client-config-data`, `apollo-mockserver` and `apollo-openapi`. If we host them in different standalone repositories, then it will be like:

* apolloconfig/apollo-java
  * apollo-client
  * apollo-core
* apolloconfig/apollo-client-config-data
* apolloconfig/apollo-mockserver
* apolloconfig/apollo-openapi

Or if we host them as different maven modules in one repository like `apollo-java`, then it will be like:

* apolloconfig/apollo-java
  * apollo-client
  * apollo-core
  * apollo-client-config-data
  * apollo-mockserver
  * apollo-openapi

Given the future maintenance cost consideration, I think hosting them as different maven modules in one repository might be better. What do you think?

**Comment by nobodyiam** [comment]
I submitted a pr #4440 to move `apollo-demo`, `scripts/docker-quick-start` and `scripts/apollo-on-kubernetes` out of main repository as the first phase. Please help to review.
* move apollo-demo to https://github.com/apolloconfig/apollo-demo-java
* move scripts/docker-quick-start to https://github.com/apolloconfig/apollo-quick-start
* move scripts/apollo-on-kubernetes to https://github.com/apolloconfig/apollo-on-kubernetes

**Comment by nobodyiam** [comment]
I submitted a pr https://github.com/apolloconfig/apollo/pull/4594 to move `apollo-core`, `apollo-client`, `apollo-mockserver`, `apollo-openapi` and `apollo-client-config-data` to [apollo-java](https://github.com/apolloconfig/apollo-java) repo. Please help to review.

* move apollo-core to https://github.com/apolloconfig/apollo-java/tree/main/apollo-core
* move apollo-client to https://github.com/apolloconfig/apollo-java/tree/main/apollo-client
* move apollo-mockserver to https://github.com/apolloconfig/apollo-java/tree/main/apollo-mockserver
* move apollo-openapi to https://github.com/apolloconfig/apollo-java/tree/main/apollo-openapi
* move apollo-client-config-data to https://github.com/apolloconfig/apollo-java/tree/main/apollo-client-config-data
